crates/red_knot_test/src/assertion.rs

```diff
@@ -0,0 +1,618 @@
+                    if !only_own_line {
+                        break;
+                    }
+                } else {
+                    break;
+                }
+            }
+            if only_own_line {
+                // The collected comments apply to the _next_ line in the code.
+                line = line.saturating_add(1);
+            }
+        } else {
+            // We have a line-trailing comment; it applies to its own line, and is not grouped.
+            collector.push(ranged_assertion.into());
+        }
+        Some(LineAssertions {
+            line,
+            assertions: collector,
+        })
+    }
+}
+
+impl std::iter::FusedIterator for LineAssertionsIterator<'_> {}
+
+/// One or more assertions referring to the same line of code.
+#[derive(Debug)]
+pub(crate) struct LineAssertions<'a> {
+    /// The line these assertions refer to.
+    ///
+    /// Not necessarily the same line the assertion comment is located on; for an own-line comment,
+    /// it's the next non-assertion line.
+    pub(crate) line: OneIndexed,
+
+    /// The assertions referring to this line.
+    pub(crate) assertions: AssertionVec<'a>,
+}
+
+static TYPE_RE: Lazy<Regex> =
+    Lazy::new(|| Regex::new(r"^#\s*Type:\s*(?<ty_display>.+?)\s*$").unwrap());
+
+static ERROR_RE: Lazy<Regex> = Lazy::new(|| {
+    Regex::new(
+        r#"^#\s*Error:(\s*(?<column>\d+))?(\s*\[(?<rule>.+?)\])?(\s*"(?<message>.+?)")?\s*$"#,
+    )
+    .unwrap()
+});
+
+/// A single diagnostic assertion comment.
+#[derive(Debug)]
+pub(crate) enum Assertion<'a> {
+    /// A `Type: ` assertion.
+    Type(TypeAssertion<'a>),
+
+    /// An `Error: ` assertion.
+    Error(ErrorAssertion<'a>),
+}
+
+impl<'a> Assertion<'a> {
+    fn from_comment(comment: &'a str) -> Option<Self> {
+        if let Some(caps) = TYPE_RE.captures(comment) {
+            Some(Self::Type(TypeAssertion {
+                ty_display: caps.name("ty_display").unwrap().as_str(),
+            }))
+        } else {
+            ERROR_RE.captures(comment).map(|caps| {
+                Self::Error(ErrorAssertion {
+                    rule: caps.name("rule").map(|m| m.as_str()),
+                    column: caps.name("column").and_then(|m| m.as_str().parse().ok()),
+                    message_contains: caps.name("message").map(|m| m.as_str()),
+                })
+            })
+        }
+    }
+}
+
+impl std::fmt::Display for Assertion<'_> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Self::Type(ta) => ta.fmt(f),
+            Self::Error(ea) => ea.fmt(f),
+        }
+    }
+}
+
+/// A `Type: ` assertion comment.
+#[derive(Debug)]
+pub(crate) struct TypeAssertion<'a> {
+    /// The type name we expect to be revealed.
+    pub(crate) ty_display: &'a str,
+}
+
+impl std::fmt::Display for TypeAssertion<'_> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "Type: {}", self.ty_display)
+    }
+}
```

Comment:
huh, is it even worth this being its own type? Maybe the enum could just be

```rs
/// A single diagnostic assertion comment.
#[derive(Debug)]
pub(crate) enum Assertion<'a> {
    /// A `Type: ` assertion.
    Type { expected_type: &'a str },

    /// An `Error: ` assertion.
    Error(ErrorAssertion<'a>),
}
```

Reply:
Yeah, that's fine, I was just keeping things more parallel, but reducing the overall code is probably better.